readthedocs/proxito/middleware.py

```diff
@@ -258,6 +168,56 @@ def add_cache_headers(self, request, response):
             # Set the key to private only if it hasn't already been set by the view.
             response.headers.setdefault('CDN-Cache-Control', 'private')
 
+    def _set_request_attributes(self, request, unresolved_domain):
+        """
+        Set attributes in the request from the unresolved domain.
+
+        - If the project was extracted from the ``X-RTD-Slug`` header,
+          we set ``request.rtdheader`` to `True`.
+        - If the project was extracted from the public domain,
+          we set ``request.subdomain`` to `True`.
+        - If the project was extracted from a custom domain,
+          we set ``request.cname`` to `True`.
+        - If the domain needs to redirect, set the canonicalize attribute accordingly.
+        """
+        # TODO: Set the unresolved domain in the request instead of each of these attributes.
+        origin = unresolved_domain.origin
+        project = unresolved_domain.project
+        if origin == OriginType.http_header:
+            request.rtdheader = True
+        elif origin == OriginType.custom_domain:
+            domain = unresolved_domain.domain
+            request.cname = True
+            request.domain = domain
+            if domain.https and not request.is_secure():
+                # Redirect HTTP -> HTTPS (302) for this custom domain.
+                log.debug("Proxito CNAME HTTPS Redirect.", domain=domain.domain)
+                request.canonicalize = constants.REDIRECT_HTTPS
+        elif origin == OriginType.external_domain:
+            request.external_domain = True
+            request.host_version_slug = unresolved_domain.external_version_slug
+        elif origin == OriginType.public_domain:
+            request.subdomain = True
+            canonical_domain = (
+                Domain.objects.filter(project=project)
+                .filter(canonical=True, https=True)
+                .exists()
+            )
+            if canonical_domain:
+                log.debug(
+                    "Proxito Public Domain -> Canonical Domain Redirect.",
+                    project_slug=project.slug,
+                )
+                request.canonicalize = constants.REDIRECT_CANONICAL_CNAME
```

Comment:
These constants should probably be an enum as well, next refactor?

Reply:
Could be, but I think we could move this logic to the docs serving view, and we won't be needing those constants anymore

https://github.com/readthedocs/readthedocs.org/pull/9982#discussion_r1098012889